### PR TITLE
[NNPA] Numerical tests for Conv2D on NNPA

### PR DESCRIPTION
**Pull request by imaihal**
This PR adds numerical tests for Conv2D on NNPA. Also, this introduces a environment variable to configure numerical tests for NNPA. 

- Add new environment variable to set data range for input data and weights.
  This PR Introduces new environment variable `{test_name}_DATARANGE` to change default data range for input data and weights.  Currently it can be used only in Conv2D on NNPA. It is provided because the data range need to be changed to avoid accuracy issue in Conv2D on NNPA. Actual range used in the numerical test for Conv2D are written in cmake file.

- Introduce new environment variable to configure test parameters for Conv2D, Gemm, LSTM, and GRU
  To include only test cases supported by NNPA, this PR provides following parameters. They need to be set environment variable, `TEST_CONFIG`
  -  Conv2D: 
     - -dim=static: only static case because dynamic height and weight dimension are not supported
     - -dilaiton=1: only default dilations (= 1) 
     - -padding=valid_upper:  only VALID and SAME_UPPER pad type.
  - Gemm: 
     - -alpha=1: alpha should be one for Matmul of zDNN library
     - -beta=1: beta should be one for Matmul of zDNN library
  - LSTM:
     - -peephole=0: LSTM of zDNN library does not support peephole tensor
  - GRU:
     - -linearBeforeReset=1: GRU of zDNN library only support LinearBeforeReset=1   
 
- Test results 
[numerical-tests.txt](https://github.com/onnx/onnx-mlir/files/9390507/numerical-tests.txt)

### CODE REVIEW

**Comment by tungld** [comment]
I feel that we have more and more environment variables for a single test. Could it be better to use a single environment variable, say TEST_CONV_OPTS="-static -d=1"? Similar for the others tests.

**Comment by imaihal** [comment]
@tungld 
> Could it be better to use a single environment variable, say TEST_CONV_OPTS="-static -d=1 -padding=valid_upper"? Similar for the other tests.

I introduced an environment variable to configure numerical tests as written in top comment in this PR.

Now ready for review. Thanks!

File: docs/AccelNNPAHowToUseAndTest.md
Comment (tungld) [inline docs/AccelNNPAHowToUseAndTest.md:34]: should `does not support` be `only supports`?

File: test/accelerators/NNPA/numerical/CMakeLists.txt
Comment (tungld) [inline test/accelerators/NNPA/numerical/CMakeLists.txt:68]: Could you please explain general settings of `data_range` and `instruction` in `docs/AccelNNPAHowToUseAndTest.md`? Is it possible to use an upperbound and lowerbound for a range instead of a single value, which is more flexible?

File: docs/AccelNNPAHowToUseAndTest.md
Comment (AlexandreEichenberger) [inline docs/AccelNNPAHowToUseAndTest.md:44]: > only LinearBeforeReset=1

Can you explain in english what is not supported? Current explanation is totological. Tx

File: docs/AccelNNPAHowToUseAndTest.md
Comment (AlexandreEichenberger) [inline docs/AccelNNPAHowToUseAndTest.md:33]: Side note, have you though of adding an `-maccel=NNPA` option that then would automatically set the current specialized options?

File: test/modellib/ConvModel.cpp
Comment (AlexandreEichenberger) [inline test/modellib/ConvModel.cpp:135]: can you redefine this one as calling the new prepareInputs with `dataRange = 1.0` please?

File: test/modellib/ModelLib.cpp
Comment (AlexandreEichenberger) [inline test/modellib/ModelLib.cpp:117]: Thanks for adding this functionality in the model.

File: test/numerical/TestConv.cpp
Comment (AlexandreEichenberger) [inline test/numerical/TestConv.cpp:30]: Are you sure you don't want to keep this printf only under DEBUG?

File: test/numerical/TestConv.cpp
Comment (AlexandreEichenberger) [inline test/numerical/TestConv.cpp:75]: It might be easier to state in the comment that `maxDilation` in an exclusive upper bound.

File: test/numerical/TestConv.cpp
Comment (AlexandreEichenberger) [inline test/numerical/TestConv.cpp:65]: The code would be more compact if you set `dimType` and `maxDilation` to their default value here, and then only redefine them when an `opts` override them. I find this style easier to read.

File: test/numerical/TestGRU.cpp
Comment (AlexandreEichenberger) [inline test/numerical/TestGRU.cpp:59]: please set the default here, and remove the 'else` clauses, tx.

File: test/numerical/TestGemm.cpp
Comment (AlexandreEichenberger) [inline test/numerical/TestGemm.cpp:89]: same here, set default here, remove `else` clauses.

**Review by AlexandreEichenberger — COMMENTED** [review]
looking good, please make small changes and it should be good to go.

File: docs/AccelNNPAHowToUseAndTest.md
Comment (imaihal) [reply docs/AccelNNPAHowToUseAndTest.md:33]: Do you mean using only `-maccel=NNPA` without using the environment variable? Yes. it should be possible.

File: test/numerical/TestConv.cpp
Comment (imaihal) [reply test/numerical/TestConv.cpp:30]: I removed it for consistency with [other tests](https://github.com/onnx/onnx-mlir/tree/main/test/numerical) The other tests don't have `DEBUG`. Even without `DEBUG`, the print statement is not output when running numerical tests using cmake command. To print it out, we need to add `--verbose` in [this line](https://github.com/onnx/onnx-mlir/blob/main/test/numerical/CMakeLists.txt#L8).

Comment (imaihal) [reply test/numerical/TestConv.cpp:30]: I will add `DEBUG` in all tests if requested.

File: docs/AccelNNPAHowToUseAndTest.md
Comment (imaihal) [reply docs/AccelNNPAHowToUseAndTest.md]: Yes. Updated. Thanks!

File: test/accelerators/NNPA/numerical/CMakeLists.txt
Comment (imaihal) [reply test/accelerators/NNPA/numerical/CMakeLists.txt]: Updated docs.

File: docs/AccelNNPAHowToUseAndTest.md
Comment (imaihal) [reply docs/AccelNNPAHowToUseAndTest.md]: Done

File: test/modellib/ConvModel.cpp
Comment (imaihal) [reply test/modellib/ConvModel.cpp:135]: Done.

File: test/numerical/TestConv.cpp
Comment (imaihal) [reply test/numerical/TestConv.cpp]: Done

File: test/numerical/TestConv.cpp
Comment (imaihal) [reply test/numerical/TestConv.cpp]: Done. Thanks!

File: test/numerical/TestGemm.cpp
Comment (imaihal) [reply test/numerical/TestGemm.cpp]: Done

File: docs/AccelNNPAHowToUseAndTest.md
Comment (AlexandreEichenberger) [reply docs/AccelNNPAHowToUseAndTest.md:33]: @imaihal 
I leave it up to you. The solution you have is clearly working, and the cmake does add the flags. It is just if we have to test one manually, we have to turn on the right flags. This is why I though that we could just use the `-maccel=NNPA` to automatically set all the options. Again, either way works, whatever you like is good.

File: test/modellib/ConvModel.cpp
Comment (AlexandreEichenberger) [reply test/modellib/ConvModel.cpp:135]: great, less redundant code

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM, will let you judge if you want to keep the code as is or simply use `-maccel=NNPA` to setup the individual internal flags (which I would definitely keep, since maybe NNPA2 will remove some of the restrictions.

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM, thanks for the great work

File: test/numerical/TestGRU.cpp
Comment (imaihal) [reply test/numerical/TestGRU.cpp]: Done. Thanks!

File: docs/AccelNNPAHowToUseAndTest.md
Comment (imaihal) [reply docs/AccelNNPAHowToUseAndTest.md:33]: @AlexandreEichenberger  I agree with you. I updated code to set them automatically when setting `--maccel=NNPA`. I kept original environment variables since they may be useful in the future.

**Pull request merged**
